packages/core/src/types.ts

```diff
@@ -73,6 +77,23 @@ export type Action<TContext, TEvent extends EventObject> =
   | ActionObject<TContext, TEvent>
   | ActionFunction<TContext, TEvent>;
 
+export type BaseAction<
+  TContext,
+  TEvent extends EventObject,
+  TAction extends BaseActionObject
+> =
+  | TAction['type']
+  | TAction
```

Comment:
This won't be type-safe - we'd have to filter out type-only actions and allow `actions: 'foo'` only for those. If an action type is defined as `{ type: 'bar', required: string }` we cannot allow `actions: 'bar'`.

Reply:
Finally made this type-safe! https://github.com/statelyai/xstate/pull/2266/files#diff-8637c68497a6e188837ab695801648e2f6f72a9831f7420c80409cc354fd1a45R407-R436